src/hotspot/share/code/compressedStream.cpp

```diff
@@ -112,3 +112,80 @@ void CompressedWriteStream::write_long(jlong value) {
   write_signed_int(low(value));
   write_signed_int(high(value));
 }
+
+
+bool CompressedSparseDataReadStream::read_zero() {
+  if (_buffer[_position] & (1 << (7 - byte_pos_))) {
+    return 0; // not a zero data
+  }
+  if (++byte_pos_ == 8) {
+    _position++;
+    byte_pos_ = 0;
+  }
+  return 1;
+}
+
+uint8_t CompressedSparseDataReadStream::read_byte_impl() {
+  uint8_t b1 = _buffer[_position] << byte_pos_;
+  uint8_t b2 = _buffer[++_position] >> (8 - byte_pos_);
+  return b1 | b2;
+}
+
+jint CompressedSparseDataReadStream::read_int() {
+  if (read_zero()) {
+    return 0;
+  }
+  // integer value encoded as a sequence of 1 to 5 bytes
+  // - the most frequent case (0 < x < 64) is encoded in one byte
+  // - the payload of the first byte is 6 bits, the payload of the following bytes is 7 bits
+  // - the most significant bit in the first byte is occupied by a zero flag
+  // - each byte has a bit indicating whether it is the last byte in the sequence
+  uint8_t b = read_byte_impl();
+  juint result = b & 0x3f;
+  for (int i = 0; (i == 0) ? (b & 0x40) : (b & 0x80); i++) {
+    b = read_byte_impl();
+    result |= ((b & 0x7f) << (6 + 7 * i));
+  }
+  return (jint)result;
+}
+
+int CompressedSparseDataWriteStream::position() {
```

Comment:
The function with a side effect looks strange to me. I see an assert in `DebugInformationRecorder::DebugInformationRecorder(OopRecorder* oop_recorder)` which uses it for checking.  So the assert can cause side affects. I am not sure it is expected.

Reply:
Storing the debug info is an iterative process. Chunks of data are compared to avoid duplication, and on some points the generated data is discarded and position is unrolled. Besides read/write stream implementation internals, DebugInformationRecorder uses raw stream data access to track the similar chunks (see [DIR_Chunk](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/debugInfoRec.cpp#L57)) and [memcpy](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/nmethod.cpp#L1969) the raw data. We have either to (1) align the data on positions where the DebugInformationRecorder splits data into chunks or to (2) take the bit position into account.

I experimented with `int position` to contain both the stream bit position in the least significant bits and stream byte position in most significant bits. For me the code becomes less readable and the performance is questionable even without the DebugInformationRecorder update:
```
-  uint8_t b1 = _buffer[_position] << _bit_pos;
-  uint8_t b2 = _buffer[++_position] >> (8 - _bit_pos);
+  uint8_t b1 = _buffer[_position >> 3] << (_position & 0x7);
+  _position += 8;
+  uint8_t b2 = _buffer[_position >> 3] >> (8 - _position & 0x7);
```
I would avoid this change and stay with current implementation. In fact, there is not much aligned positions within the data. And `assert(_stream->position() > serialized_null, "sanity");` (thanks for noticing that!) in the constructor makes no problem because data is aligned at the beginning of the stream.

Reply:
If we introduce `flush()` we can explicitly call it before any accesses to the internal buffer: `buffer()` and `position()`. We will add asserts to `buffer()` and `position()` checking whether there is unflushed data. I always prefer being explicit.